File: etl/utils/preprocesser.py
from typing import List
import datetime
from tiktoken import Tokenizer

CHUNK_SIZE = 4000
OVERLAP = 200
INPUT_FORMAT = '%d/%m/%Y'
OUTPUT_FORMAT = '%Y-%m-%d'

def chop(text: str, chunk_size: int = CHUNK_SIZE, overlap: int = OVERLAP) -> List[str]:
    """
    Takes a text as input and returns a list
    of chunks of max size chunk_size. GPT tokens
    are used to determine the chunk size.
    """
    tokenizer = Tokenizer()

    tokens = tokenizer.tokenize(text)
    num_tokens = len(tokens)
    
    chunks = []
    for i in range(0, num_tokens, chunk_size - overlap):
        chunk = tokens[i:i + chunk_size]
        chunks.append(''.join(chunk))
    
    return chunks

def transform_date(date: str, input_format: str = INPUT_FORMAT, output_format: str = OUTPUT_FORMAT) -> str:
    """
    Transform str date from input format to
    output format.
    """
    try:
        date = datetime.datetime.strptime(date, input_format)
        date = date.strftime(output_format)
    except ValueError as e:
        print(f"Error occurred: {e}")
        return None

    return date

File: internal/preprocesser/preprocesser.go
package preprocesser

import (
	"borai/internal/loader"
	"time"
)

func Chop(text string, chunkSize, overlap int) []string {

	var chunks []string
	for i := 0; i < len(text); i += (chunkSize - overlap) * 4 {
		end := i + chunkSize*4
		if end > len(text) {
			end = len(text)
		}
		chunks = append(chunks, text[i:end])
	}
	return chunks

}

func TransformDate(date, inputFormat, outputFormat string) (string, error) {
	t, err := time.Parse(inputFormat, date)
	if err != nil {
		return "", err
	}
	return t.Format(outputFormat), nil
}

func Sort(data []loader.Publication, date time.Time) []loader.Publication {
	var sorted []loader.Publication
	dateStr := date.Format("2006-01-02")
	for _, p := range data {
		if p.Date == dateStr {
			sorted = append(sorted, p)
		}
	}

	return sorted

}
